From df38f25e5c3789e94dd29b8157aa5ec0ba3f8f50 Mon Sep 17 00:00:00 2001
From: Li Jun <B47624@freescale.com>
Date: Sat, 27 Sep 2014 20:10:43 +0800
Subject: [PATCH 0120/5242] MLK-9637-3 usb: chipidea: udc: test: handle
 otg_srp_reqd set feature request

commit  0a8886f6ad3db52292c1a7a6ffe24eecde11cf39 from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds otg_srp_reqd set feature request from A host for chipidea
udc driver.

Acked-by: Peter Chen <peter.chen@freescale.com>
Signed-off-by: Li Jun <b47624@freescale.com>
(cherry picked from commit 5edca27c518e8a69accb429c946c19243e156656)
(cherry picked from commit 14ac2f1f74ce0b02e03296472e7d170c45a31f29)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/udc.c |   16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/usb/chipidea/udc.c b/drivers/usb/chipidea/udc.c
index 63b740a..5e80b80 100644
--- a/drivers/usb/chipidea/udc.c
+++ b/drivers/usb/chipidea/udc.c
@@ -749,6 +749,9 @@ static void isr_reset_handler(struct ci_hdrc *ci)
 {
 	int retval;
 
+	if (ci_otg_is_fsm_mode(ci))
+		ci->fsm.otg_srp_reqd = 0;
+
 	spin_unlock(&ci->lock);
 	if (ci->gadget.speed != USB_SPEED_UNKNOWN)
 		usb_gadget_udc_reset(&ci->gadget, ci->driver);
@@ -1014,6 +1017,16 @@ static int otg_a_alt_hnp_support(struct ci_hdrc *ci)
 	return isr_setup_status_phase(ci);
 }
 
+static int otg_srp_reqd(struct ci_hdrc *ci)
+{
+	if (ci_otg_is_fsm_mode(ci)) {
+		ci->fsm.otg_srp_reqd = 1;
+		return isr_setup_status_phase(ci);
+	} else {
+		return -ENOTSUPP;
+	}
+}
+
 /**
  * isr_setup_packet_handler: setup packet handler
  * @ci: UDC descriptor
@@ -1137,6 +1150,9 @@ static void isr_setup_packet_handler(struct ci_hdrc *ci)
 					err = isr_setup_status_phase(
 							ci);
 					break;
+				case TEST_OTG_SRP_REQD:
+					err = otg_srp_reqd(ci);
+					break;
 				default:
 					break;
 				}
-- 
1.7.9.5

